### PR TITLE
migrate animation to nullsafety

### PR DESCRIPTION
**Pull request by a14n**
## Description

NNBD migration for animation

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/animation/animation_controller.dart [hunk 1]
```diff
@@ -414,22 +413,22 @@ class AnimationController extends Animation<double>
   /// and the most recent tick of the animation.
   ///
   /// If the controller is not animating, the last elapsed duration is null.
-  Duration get lastElapsedDuration => _lastElapsedDuration;
-  Duration _lastElapsedDuration;
+  Duration get lastElapsedDuration => _lastElapsedDuration!;
```
Comment (goderbauer) [inline packages/flutter/lib/src/animation/animation_controller.dart:416]: This is conflicting with the doc comment above it.

Comment (a14n) [reply packages/flutter/lib/src/animation/animation_controller.dart]: Fixed

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by datayeah** [comment]
i now get "The named parameter 'vsync' isn't defined." when trying to use an AnimationController.
`AnimationController(
        duration: const Duration(milliseconds: 800),vsync: this)`

**Comment by a14n** [comment]
`vsync` param is still here https://github.com/flutter/flutter/pull/62485/files#diff-82a88b0b4aa73ca8c85c9fc068be95b6R242 .
This issue may be an issue regarding null-safety handling in your project. Is there a public repo reproducing the issue?

**Comment by lazylazyllama** [comment]
Its probably because the analyzer does not understand the "required" keyword? which was @required before?

**Comment by datayeah** [comment]
@a14n My IDE (Android Studio 4.0.1) tells me that vsync does not exist and marks this as an error. It suggests adding the vsync parameter with label "TickerProvider:" instead of "vsync:". Fun fact: It compiles when using "vsync:", and it fails to compile if i use the suggested "TickerProvider:" label. I ran "flutter clean" but it doesnt change. I will try to investigate further...

**Comment by datayeah** [comment]
> Its probably because the analyzer does not understand the "required" keyword? which was @required before?

yes, if i temporarily change it back to "@required" in the flutter sources, the vsync param is "detected" by the IDE, but then it crashes due to other changes that were made.

**Comment by a14n** [comment]
Can you try to run `flutter analyze` directly?

**Comment by datayeah** [comment]
No issues found! (ran in 12.7s;)

**Comment by a14n** [comment]
So it looks like your IDE use a version of Dart that is not the one packaged with Flutter.

**Comment by datayeah** [comment]
Dart SDK version: 2.10.0-1.0.dev.flutter-24c7666def (be) (Wed Jul 22 13:22:07 2020 +0000) on "macos_x64"

**Comment by datayeah** [comment]
Okay, to complete this and (maybe) help others: 
The issue disappeared after restarting the IDE. 
Note to self: Restart the IDE after "flutter upgrade".

**Comment by thisisgit** [comment]
@a14n Came from https://github.com/flutter/flutter/issues/62574 and just wondering the difference between `required` vs. `@required`? 
Is it something that we all need to be prepared for the future release?

**Comment by a14n** [comment]
`required` has the same meaning in nullsafety as `@required` before nullsafety. See https://dart.dev/null-safety/understanding-null-safety#required-named-parameters

>Is it something that we all need to be prepared for the future release?

As nullsafety is available behind an experimental flag I would not recommand to migrate your code for now.

**Pull request merged**